### PR TITLE
Update queries to return only available variants

### PR DESCRIPTION
**Pull request by d-wysocki**
- For users return only available products and variants with price
- For staff return all products and variants
- updated filtering by `price`, `attributes`, `stock` to return available products and variants with price for users
- updated sort by `price` to return available products and variants with price for users

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-2546 Update queries to return only available variants](https://app.clickup.com/t/2549495/SALEOR-2546)

File: saleor/graphql/core/tests/test_graphql.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def test_real_query(user_api_client, product, channel_USD):
             __typename
         }
         products(first: $first, sortBy: $sortBy, filter: {categories: [$categoryId],
-            attributes: $attributesFilter}, channel: $channel) {
+            attributes: $attributesFilter,channel: $channel}, channel: $channel) {
```
Comment (IKarbowiak) [inline saleor/graphql/core/tests/test_graphql.py:89]: Not needed change.

File: saleor/graphql/product/dataloaders/products.py [hunk 2]
```diff
@@ -166,6 +166,35 @@ def batch_load(self, keys):
         return [variant_map.get(product_id, []) for product_id in keys]
 
 
+class AvailableProductVariantsByProductVariantId(
```
Comment (IKarbowiak) [inline saleor/graphql/product/dataloaders/products.py:169]: IMO, for consistency it should be:
```suggestion
class AvailableProductVariantsByProductVariantIdAndChannel
```

File: saleor/graphql/product/tests/test_product.py [hunk 3]
```diff
@@ -1761,6 +1768,50 @@ def test_products_query_with_filter(
...
+        "filter": {"price": {"gte": 9, "lte": 31}, "channel": channel_USD.slug},
+        "channel": channel_USD.slug,
+    }
+    user_api_client.user.user_permissions.add(permission_manage_products)
+    response = user_api_client.post_graphql(query_products_with_filter, variables)
+    content = get_graphql_content(response)
+    products = content["data"]["products"]["edges"]
+
+    assert len(products) == 2
+
```
Comment (IKarbowiak) [inline saleor/graphql/product/tests/test_product.py:1813]: IMO we should also check if the product with no `price_amount` isn't in the returned products list.

File: saleor/graphql/product/tests/test_product.py [hunk 4]
```diff
@@ -1786,27 +1837,70 @@ def test_products_query_with_filter_search_by_sku(
...
+    }
     staff_api_client.user.user_permissions.add(permission_manage_products)
     response = staff_api_client.post_graphql(query_products_with_filter, variables)
     content = get_graphql_content(response)
-    product_id = graphene.Node.to_global_id("Product", product.id)
     products = content["data"]["products"]["edges"]
 
-    assert len(products) == 1
+    assert len(products) == 3
+
```
Comment (IKarbowiak) [inline saleor/graphql/product/tests/test_product.py:1868]: Same here.

File: saleor/graphql/product/tests/test_product.py [hunk 5]
```diff
@@ -2065,10 +2162,88 @@ def test_filter_products_by_wrong_attributes(user_api_client, product, channel_U
...
+    product_list[0].variants.first().channel_listings.filter(
+        channel=channel_USD
+    ).update(price_amount=None)
+
+    response = user_api_client.post_graphql(query, variables)
+    content = get_graphql_content(response)
+    products = content["data"]["products"]["edges"]
+
+    assert len(products) == 2
+
```
Comment (IKarbowiak) [inline saleor/graphql/product/tests/test_product.py:2202]: Same here, with variants. 👮‍♀️

File: saleor/graphql/product/tests/test_product.py (hunk 4 above)
Comment (d-wysocki) [reply saleor/graphql/product/tests/test_product.py:1868]: In this case we are receiving all products with and without price.

File: saleor/graphql/product/types/products.py [hunk 6]
```diff
@@ -727,15 +732,42 @@ def resolve_images(root: ChannelContext[models.Product], info, **_kwargs):
...
     def resolve_variants(root: ChannelContext[models.Product], info, **_kwargs):
+        requestor = get_user_or_app_from_context(info.context)
+        is_staff = requestor_is_staff_member_or_app(requestor)
+
+        def return_available_variants_in_channel(variants):
+            if is_staff:
+                return variants
+
+            variants_id = (variant.id for variant in variants)
+            variant_channel_listings = (
```
Comment (fowczarek) [inline saleor/graphql/product/types/products.py:743]: `variant_channel_listings` - This dataloader doesn't return ProductVariantChannelListing

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/product/filters.py [hunk 7]
```diff
@@ -61,7 +61,9 @@ def _clean_product_attributes_filter_input(
 T_PRODUCT_FILTER_QUERIES = Dict[int, Iterable[int]]
 
 
-def filter_products_by_attributes_values(qs, queries: T_PRODUCT_FILTER_QUERIES):
+def filter_products_by_attributes_values(
+    qs, queries: T_PRODUCT_FILTER_QUERIES, channel_slug
```
[the remark this answers is not in the document]
Comment (d-wysocki) [reply saleor/graphql/product/filters.py]: My mistake, this left from other solution.

File: saleor/graphql/product/dataloaders/products.py [hunk 8]
```diff
@@ -166,6 +166,35 @@ def batch_load(self, keys):
...
+
+    def batch_load_available(
+        self, channel: str, variant_ids: Iterable[int]
+    ) -> Iterable[int]:
+        filter = {
+            f"channel__{self.field}": channel,
+            "variant_id__in": variant_ids,
+            "price_amount__isnull": False,
+        }
+        variant_channel_listings = ProductVariantChannelListing.objects.filter(**filter)
```
[the remark this answers is not in the document]
Comment (d-wysocki) [reply saleor/graphql/product/dataloaders/products.py:190]: I will try to refactor this dataloader.

**Review by maarcingebala — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Pull request merged**